File: src/OllamaFlow.Core/Models/Backend.cs
namespace OllamaFlow.Core
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Net.Http;
    using System.Text;
    using System.Text.Json.Serialization;
    using System.Threading;
    using System.Threading.Tasks;
    using OllamaFlow.Core.Serialization;
    using OllamaFlow.Core.Enums;

    /// <summary>
    /// Origin server.
    /// </summary>
    public class Backend
    {
        #region Public-Members

        /// <summary>
        /// Serializer.
        /// </summary>
        [JsonIgnore]
        public static Serializer Serializer = new Serializer();

        /// <summary>
        /// Unique identifier for this origin server.
        /// </summary>
        public string Identifier { get; set; } = null;

        /// <summary>
        /// Name for this origin server.
        /// </summary>
        public string Name { get; set; } = null;

        /// <summary>
        /// Hostname.
        /// </summary>
        public string Hostname
        {
            get
            {
                return _Hostname;
            }
            set
            {
                if (String.IsNullOrEmpty(value)) throw new ArgumentNullException(nameof(Hostname));
                _Hostname = value;
            }
        }

        /// <summary>
        /// TCP port.
        /// </summary>
        public int Port
        {
            get
            {
                return _Port;
            }
            set
            {
                if (value < 0 || value > 65535) throw new ArgumentOutOfRangeException(nameof(Port));
                _Port = value;
            }
        }

        /// <summary>
        /// Enable or disable SSL.
        /// </summary>
        public bool Ssl { get; set; } = false;

        /// <summary>
        /// Number of consecutive failed health checks before marking a server as unhealthy.
        /// Default is 2.
        /// </summary>
        public int UnhealthyThreshold
        {
            get
            {
                return _UnhealthyThreshold;
            }
            set
            {
                if (value < 1) throw new ArgumentOutOfRangeException(nameof(UnhealthyThreshold));
                _UnhealthyThreshold = value;
            }
        }

        /// <summary>
        /// Number of consecutive successful health checks before marking a server as healthy.
        /// Default is 2.
        /// </summary>
        public int HealthyThreshold
        {
            get
            {
                return _HealthyThreshold;
            }
            set
            {
                if (value < 1) throw new ArgumentOutOfRangeException(nameof(HealthyThreshold));
                _HealthyThreshold = value;
            }
        }

        /// <summary>
        /// HTTP method to use when performing a healthcheck.
        /// Default is GET.
        /// </summary>
        public HttpMethod HealthCheckMethod
        {
            get
            {
                return _HealthCheckMethod;
            }
            set
            {
                if (value == null) throw new ArgumentNullException(nameof(HealthCheckMethod));
                _HealthCheckMethod = value;
            }
        }

        /// <summary>
        /// URL to use when performing a healthcheck.
        /// Default is /.
        /// </summary>
        public string HealthCheckUrl
        {
            get
            {
                return _HealthCheckUrl;
            }
            set
            {
                if (String.IsNullOrEmpty(value)) value = "/";
                _HealthCheckUrl = value;
            }
        }

        /// <summary>
        /// Maximum number of parallel requests to this backend.
        /// Default is 4.
        /// </summary>
        public int MaxParallelRequests
        {
            get
            {
                return _MaxParallelRequests;
            }
            set
            {
                if (value < 1) throw new ArgumentOutOfRangeException(nameof(MaxParallelRequests));
                _MaxParallelRequests = value;
            }
        }

        /// <summary>
        /// Threshold at which 429 rate limit responses are sent.
        /// Default is 10.
        /// </summary>
        public int RateLimitRequestsThreshold
        {
            get
            {
                return _RateLimitRequestsThreshold;
            }
            set
            {
                if (value < 1) throw new ArgumentOutOfRangeException(nameof(RateLimitRequestsThreshold));
                _RateLimitRequestsThreshold = value;
            }
        }

        /// <summary>
        /// URL prefix.
        /// </summary>
        [JsonIgnore]
        public string UrlPrefix
        {
            get
            {
                return (Ssl ? "https://" : "http://") + Hostname + ":" + Port;
            }
        }

        /// <summary>
        /// Boolean indicating if the full request should be logged.
        /// </summary>
        public bool LogRequestFull { get; set; } = false;

        /// <summary>
        /// Boolean indicating if the request body should be logged.
        /// </summary>
        public bool LogRequestBody { get; set; } = false;

        /// <summary>
        /// Boolean indicating if the response body should be logged.
        /// </summary>
        public bool LogResponseBody { get; set; } = false;

        /// <summary>
        /// API format supported by this backend.
        /// Default is Ollama.
        /// </summary>
        public ApiFormatEnum ApiFormat { get; set; } = ApiFormatEnum.Ollama;

        /// <summary>
        /// Boolean indicating if the object is active or not.
        /// </summary>
        public bool Active { get; set; } = true;

        /// <summary>
        /// Creation timestamp, in UTC time.
        /// </summary>
        public DateTime CreatedUtc { get; set; } = DateTime.UtcNow;

        /// <summary>
        /// Last update timestamp, in UTC time.
        /// </summary>
        public DateTime LastUpdateUtc { get; set; } = DateTime.UtcNow;

        /// <summary>
        /// Timestamp at which the backend was seen as healthy, in UTC time.
        /// </summary>
        public DateTime? HealthySinceUtc { get; set; }

        /// <summary>
        /// Timestamp at which the backend was seen as unhealthy, in UTC time.
        /// </summary>
        public DateTime? UnhealthySinceUtc { get; set; }

        /// <summary>
        /// Uptime for the backend.
        /// </summary>
        public TimeSpan? Uptime
        {
            get
            {
                if (HealthySinceUtc != null) return DateTime.UtcNow - HealthySinceUtc.Value;
                return null;
            }
        }

        /// <summary>
        /// Downtime for the backend.
        /// </summary>
        public TimeSpan? Downtime
        {
            get
            {
                if (UnhealthySinceUtc != null) return DateTime.UtcNow - UnhealthySinceUtc.Value;
                return null;
            }
        }

        /// <summary>
        /// Number of active requests.
        /// </summary>
        public int ActiveRequests
        {
            get
            {
                return _ActiveRequests;
            }
            set
            {
                if (value < 0) throw new ArgumentOutOfRangeException(nameof(ActiveRequests));
                _ActiveRequests = value;
            }
        }

        /// <summary>
        /// Boolean indicating whether or not the backend was chosen due to stickiness.
        /// </summary>
        public bool IsSticky { get; set; } = false;

        #endregion

        #region Internal-Members

        internal readonly object Lock = new object();
        internal bool Healthy = false;
        internal int HealthCheckSuccess = 0;
        internal int HealthCheckFailure = 0;
        internal bool ModelsDiscovered = false;
        internal int _ActiveRequests = 0;
        internal int _PendingRequests = 0;
        internal List<string> Models
        {
            get
            {
                return _Models;
            }
            set
            {
                if (value == null) value = new List<string>();
                _Models = value;
            }
        }
        internal SemaphoreSlim Semaphore
        {
            get
            {
                if (_Semaphore == null) _Semaphore = new SemaphoreSlim(_MaxParallelRequests, _MaxParallelRequests);
                return _Semaphore;
            }
            set
            {
                _Semaphore = value;
            }
        }

        #endregion

        #region Private-Members

        private string _Hostname = "localhost";
        private int _Port = 8000;
        private int _UnhealthyThreshold = 2;
        private int _HealthyThreshold = 2;
        private int _MaxParallelRequests = 4;
        private int _RateLimitRequestsThreshold = 10;
        private HttpMethod _HealthCheckMethod = HttpMethod.Get;
        private string _HealthCheckUrl = "/";
        private List<string> _Models = new List<string>();
        private SemaphoreSlim _Semaphore = null;

        #endregion

        #region Constructors-and-Factories

        /// <summary>
        /// Backend server.
        /// </summary>
        public Backend()
        {

        }

        #endregion

        #region Public-Methods

        #endregion

        #region Private-Methods

        #endregion
    }
}

File: sdk/python/src/ollamaflow_sdk/models/backend.py
"""Backend models for OllamaFlow SDK."""

from datetime import datetime
from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, Field


class ApiFormat(str, Enum):
    """API format supported by a backend."""

    OLLAMA = "Ollama"
    OPENAI = "OpenAI"


class Backend(BaseModel):
    """Represents a backend AI inference instance in OllamaFlow."""

    identifier: Optional[str] = Field(default=None, alias="Identifier")
    """Unique identifier for this backend."""

    name: Optional[str] = Field(default=None, alias="Name")
    """Human-readable name for this backend."""

    hostname: str = Field(default="localhost", alias="Hostname")
    """Hostname or IP address of the backend."""

    port: int = Field(default=11434, alias="Port", ge=0, le=65535)
    """TCP port of the backend."""

    ssl: bool = Field(default=False, alias="Ssl")
    """Whether to use SSL/TLS for connections."""

    unhealthy_threshold: int = Field(default=2, alias="UnhealthyThreshold", ge=1)
    """Number of consecutive failed health checks before marking as unhealthy."""

    healthy_threshold: int = Field(default=2, alias="HealthyThreshold", ge=1)
    """Number of consecutive successful health checks before marking as healthy."""

    health_check_method: str = Field(default="GET", alias="HealthCheckMethod")
    """HTTP method to use for health checks (GET or HEAD)."""

    health_check_url: str = Field(default="/", alias="HealthCheckUrl")
    """URL path to use for health checks."""

    max_parallel_requests: int = Field(default=4, alias="MaxParallelRequests", ge=1)
    """Maximum number of parallel requests allowed to this backend."""

    rate_limit_requests_threshold: int = Field(
        default=10, alias="RateLimitRequestsThreshold", ge=1
    )
    """Threshold at which rate limiting begins."""

    log_request_full: bool = Field(default=False, alias="LogRequestFull")
    """Whether to log full request details."""

    log_request_body: bool = Field(default=False, alias="LogRequestBody")
    """Whether to log request bodies."""

    log_response_body: bool = Field(default=False, alias="LogResponseBody")
    """Whether to log response bodies."""

    api_format: ApiFormat = Field(default=ApiFormat.OLLAMA, alias="ApiFormat")
    """API format supported by this backend."""

    labels: list[str] = Field(default_factory=list, alias="Labels")
    """Labels for backend selection in load balancing."""

    pinned_embeddings_properties: dict[str, Any] = Field(
        default_factory=dict, alias="PinnedEmbeddingsProperties"
    )
    """Properties to apply to all embeddings requests."""

    pinned_completions_properties: dict[str, Any] = Field(
        default_factory=dict, alias="PinnedCompletionsProperties"
    )
    """Properties to apply to all completions requests."""

    allow_embeddings: bool = Field(default=True, alias="AllowEmbeddings")
    """Whether embeddings requests are allowed."""

    allow_completions: bool = Field(default=True, alias="AllowCompletions")
    """Whether completions requests are allowed."""

    bearer_token: Optional[str] = Field(default=None, alias="BearerToken")
    """Bearer token for authentication with this backend."""

    querystring: Optional[str] = Field(default=None, alias="Querystring")
    """Query string to append to backend URLs."""

    headers: dict[str, str] = Field(default_factory=dict, alias="Headers")
    """Custom headers to include in requests to this backend."""

    active: bool = Field(default=True, alias="Active")
    """Whether this backend is active."""

    created_utc: Optional[datetime] = Field(default=None, alias="CreatedUtc")
    """Timestamp when this backend was created."""

    last_update_utc: Optional[datetime] = Field(default=None, alias="LastUpdateUtc")
    """Timestamp when this backend was last updated."""

    active_requests: int = Field(default=0, alias="ActiveRequests", ge=0)
    """Number of currently active requests."""

    is_sticky: bool = Field(default=False, alias="IsSticky")
    """Whether this backend was selected due to session stickiness."""

    model_config = {"populate_by_name": True}


class BackendHealth(Backend):
    """Backend with additional health status information."""

    healthy_since_utc: Optional[datetime] = Field(default=None, alias="HealthySinceUtc")
    """Timestamp when this backend was last seen as healthy."""

    unhealthy_since_utc: Optional[datetime] = Field(default=None, alias="UnhealthySinceUtc")
    """Timestamp when this backend was last seen as unhealthy."""

    uptime: Optional[str] = Field(default=None, alias="Uptime")
    """Duration the backend has been healthy."""

    downtime: Optional[str] = Field(default=None, alias="Downtime")
    """Duration the backend has been unhealthy."""
